aptos-move/aptos-vm/src/verifier/transaction_arg_validation.rs

```diff
@@ -558,3 +557,196 @@ fn read_n_bytes(n: usize, src: &mut Cursor<&[u8]>, dest: &mut Vec<u8>) -> Result
+    module_storage: &impl ModuleStorage,
+    module_id: &ModuleId,
+    function_name: &IdentStr,
+    expected_return_ty: &Type,
+) -> VMResult<LoadedFunction> {
+    // Here, we do not charge gas for module loading due to invoking a constructor function. This
+    // is safe to do because all constructor functions are located at 0x1 (special address) and so
+    // should not be charged.
+    if !module_id.address().is_special() {
+        let msg = format!(
+            "Constructor function {}::{}::{} has a non-special address!",
+            module_id.address(),
+            module_id.name(),
+            function_name
+        );
+        let err = PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
+            .with_message(msg)
+            .finish(Location::Undefined);
+        return Err(err);
+    }
+    let (module, function) = module_storage.fetch_function_definition(
+        module_id.address(),
+        module_id.name(),
+        function_name,
+    )?;
+
+    if function.return_tys().len() != 1 {
+        // For functions that are marked constructor this should not happen.
+        return Err(PartialVMError::new(StatusCode::ABORTED).finish(Location::Undefined));
+    }
+
+    let mut map = TypeParamMap::default();
+    if !map.match_ty(&function.return_tys()[0], expected_return_ty) {
+        // For functions that are marked constructor this should not happen.
+        return Err(
+            PartialVMError::new(StatusCode::INVALID_MAIN_FUNCTION_SIGNATURE)
+                .finish(Location::Undefined),
+        );
+    }
+
+    // Construct the type arguments from the match.
+    let num_ty_args = function.ty_param_abilities().len();
+    let mut ty_args = Vec::with_capacity(num_ty_args);
+    for i in 0..num_ty_args {
+        ty_args.push(map.get_ty_param(i as u16)?);
+    }
+
+    Type::verify_ty_arg_abilities(function.ty_param_abilities(), &ty_args)
+        .map_err(|e| e.finish(Location::Module(module_id.clone())))?;
+
+    Ok(LoadedFunction {
+        owner: LoadedFunctionOwner::Module(module),
+        ty_args,
+        function,
+    })
+}
+
+#[derive(Default)]
+struct TypeParamMap<'a> {
+    map: BTreeMap<u16, &'a Type>,
+}
+
+impl<'a> TypeParamMap<'a> {
+    fn get_ty_param(&self, idx: u16) -> VMResult<Type> {
+        self.map.get(&idx).map(|ty| (*ty).clone()).ok_or_else(|| {
+            PartialVMError::new(StatusCode::INVALID_MAIN_FUNCTION_SIGNATURE)
+                .finish(Location::Undefined)
+        })
+    }
+
+    /// Matches the actual returned type to the expected type, binding any type args to the necessary
+    /// type as stored in the map. The expected type must be a concrete type (no [Type::TyParam]).
+    /// Returns true if a successful match is made.
+    // TODO: is this really needed in presence of paranoid mode? This does a deep structural
+    //       comparison and is expensive.
+    fn match_ty(&mut self, ty: &Type, expected: &'a Type) -> bool {
```

Comment:
I wonder whether this should really go into the runtime type module, including the TypeParamMap.

Reply:
moved to runtime type module